src/AvTranscoder/filter/FilterGraph.cpp

```diff
@@ -16,6 +15,123 @@ extern "C" {
 namespace avtranscoder
 {
 
+
+/******************
+
+    AudioFrameBuffer
+
+ ******************/
+
+AudioFrameBuffer::AudioFrameBuffer(const AudioFrameDesc& audioFrameDesc)
+    : _audioFrameDesc(audioFrameDesc)
+    , _frameQueue()
+    , _totalDataSize(0)
+    , _positionInFrontFrame(0)
+{
+}
+
+AudioFrameBuffer::~AudioFrameBuffer()
+{
+    for(size_t i = 0; i < _frameQueue.size(); ++i)
+        popFrame();
+}
+
+size_t AudioFrameBuffer::getBytesPerSample()
+{
+    return av_get_bytes_per_sample(_audioFrameDesc._sampleFormat);
+}
+
+void AudioFrameBuffer::addFrame(IFrame* frame)
+{
+    LOG_DEBUG("Add a new " << frame->getDataSize() << " bytes frame to frame buffer. New buffer size: " << _frameQueue.size() + 1);
+    // Copy the input frame to store it into the queue
+    AudioFrame* newAudioFrame = new AudioFrame(_audioFrameDesc, false);
+    const size_t expectedNbSamples = frame->getDataSize() / (newAudioFrame->getNbChannels() * newAudioFrame->getBytesPerSample());
+    newAudioFrame->setNbSamplesPerChannel(expectedNbSamples);
+    newAudioFrame->allocateData();
+    newAudioFrame->copyData(*frame);
+
+    _totalDataSize += newAudioFrame->getDataSize();
+    _frameQueue.push(newAudioFrame);
+}
+
+void AudioFrameBuffer::popFrame()
+{
+    _frameQueue.pop();
+    LOG_DEBUG("Pop frame from buffer. Remaining frames in buffer: " << _frameQueue.size());
+}
+
+IFrame* AudioFrameBuffer::getFrame(const size_t size)
+{
+    LOG_DEBUG("Get a " << size << " bytes frame from a " << _totalDataSize << " bytes frame buffer");
+    IFrame* next = _frameQueue.front();
+    const size_t nextFrameSize = next->getDataSize();
+
+    // If no expected size, or if the expected size equals the front frame of the queue (with no offset)
+    if(size == 0 || (size == nextFrameSize && _positionInFrontFrame == 0))
+    {
+        // Directly return the front frame of the queue
+        _totalDataSize -= nextFrameSize;
+        popFrame();
+        return next;
+    }
+
+    // Create a new frame
+    AudioFrame* newAudioFrame = new AudioFrame(_audioFrameDesc, false);
+    const size_t expectedNbSamples = size / (newAudioFrame->getNbChannels() * newAudioFrame->getBytesPerSample());
+    newAudioFrame->setNbSamplesPerChannel(expectedNbSamples);
+    newAudioFrame->allocateData();
+
+    // Concatenate frames data
+    size_t extractedDataSize = 0;
+    unsigned char* outputData = new unsigned char[size];
```

Comment:
Is this type mean that the supported data type is only 1byte per sample?

Reply:
No, it is actually the raw data in bytes, without considering the sample format or whatever.